### ISSUE TITLE
If AC off turning on throws error KeyError: '@AC_MAIN_OPERATION_ALL_ON_W

### ISSUE THREAD

**Issue by petarvucetin** [sampsyo/wideq#83]
If AC is off it cannot be turned back on. ([Issue from hass-smarthinq](https://github.com/sampsyo/hass-smartthinq/issues/45))

**Model:** LG 115V Dual Inverter Technology Portable Air Conditioner
**ASIN:** B07NC2QM6V
**Item model number:**  LP1419IVSM

**Output from example.py**:
d2806ee0-xxxx-xxxx-xxxx-7440beb80834: Office AC (AC POT_056905_WW)

```
 [master] $ python3 example.py ac-mon d2806ee0-xxxx-xxxx-xxxx-7440beb80834
off; COOL; cur 68°F; cfg 75°F; fan speed LOW
off; COOL; cur 68°F; cfg 75°F; fan speed LOW
off; COOL; cur 68°F; cfg 75°F; fan speed LOW
off; COOL; cur 68°F; cfg 75°F; fan speed LOW
```
```
 [master !] $ python3 example.py ac-config d2806ee0-xxxx-xxxx-xxxx-7440beb80834
************DEBUG DUMP DATA *********
b'{"UseTime":"0","ChangePeriod":"720","ChangeDate":"00000000"}'
************DEBUG DUMP DATA END *********
{'UseTime': '0', 'ChangePeriod': '720', 'ChangeDate': '00000000'}
************DEBUG DUMP DATA *********
b'{"RemainTime":"98","ChangePeriod":"250"}'
************DEBUG DUMP DATA END *********
{'RemainTime': '98', 'ChangePeriod': '250'}
************DEBUG DUMP DATA *********
b'{"Day":"0","Week":"0","Month":"0"}'
************DEBUG DUMP DATA END *********
{'Day': '0', 'Week': '0', 'Month': '0'}
************DEBUG DUMP DATA *********
b'{"InOutInstantPower":"50"}'
************DEBUG DUMP DATA END *********
50  watts
************DEBUG DUMP DATA *********
b'{"OutTotalInstantPower":"0"}'
************DEBUG DUMP DATA END *********
0  watts
0
True
************DEBUG DUMP DATA *********
b'{[{"No":"1","Cfg":"1","State":"0"},{"No":"2","Cfg":"0","State":"0"},{"No":"3","Cfg":"0","State":"0"},{"No":"4","Cfg":"0","State":"0"},{"No":"5","Cfg":"0","State":"0"},{"No":"6","Cfg":"0","State":"0"},{"No":"7","Cfg":"0","State":"0"},{"No":"8","Cfg":"0","State":"0"}]}'
************DEBUG DUMP DATA END *********
Traceback (most recent call last):
  File "example.py", line 269, in <module>
    main()
  File "example.py", line 265, in main
    example(args.country, args.language, args.verbose, args.cmd, args.args)
  File "example.py", line 206, in example
    example_command(client, cmd, args)
  File "example.py", line 175, in example_command
    func(client, *args)
  File "example.py", line 156, in ac_config
    print(ac.get_zones())
  File "/mnt/e/Development/github/wideq/wideq/ac.py", line 198, in get_zones
    return self._get_config('DuctZone')
  File "/mnt/e/Development/github/wideq/wideq/client.py", line 440, in _get_config
    return json.loads(base64.b64decode(data).decode('utf8'))
  File "/usr/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.6/json/decoder.py", line 355, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)
```
```
[master] $ python3 example.py turn d2806ee0-xxxx-xxxx-xxxx-7440beb80834 on
Traceback (most recent call last):
  File "example.py", line 269, in <module>
    main()
  File "example.py", line 265, in main
    example(args.country, args.language, args.verbose, args.cmd, args.args)
  File "example.py", line 206, in example
    example_command(client, cmd, args)
  File "example.py", line 175, in example_command
    func(client, *args)
  File "example.py", line 144, in turn
    ac.set_on(on_off == 'on')
  File "/mnt/e/Development/github/wideq/wideq/ac.py", line 233, in set_on
    op_value = self.model.enum_value('Operation', op.value)
  File "/mnt/e/Development/github/wideq/wideq/client.py", line 345, in enum_value
    return options_inv[name]
KeyError: '@AC_MAIN_OPERATION_ALL_ON_W'

```

**Comment by sampsyo** [comment]
Got it; thanks! I think the most useful thing here might be the device details from the JSON file that example.py creates. Can you remove your tokens from that file (which are authentication credentials and shouldn't be shared) and include it?

**Comment by petarvucetin** [comment]
Attached.
[wideq_state.zip](https://github.com/sampsyo/wideq/files/4176515/wideq_state.zip)

**Comment by sampsyo** [comment]
Thanks! The culprit is here:

```json
"Operation": {
  "_comment": "운전On/Off",
  "type": "Enum",
  "default": "0",
  "option": {
    "0": "@AC_MAIN_OPERATION_OFF_W",
    "1": "@AC_MAIN_OPERATION_RIGHT_ON_W"
  }
},
```

Namely, the "on" state for your model is `@AC_MAIN_OPERATION_RIGHT_ON_W`. This used to be what we used for "on," but #57 changed the "on/off" states to have separate states for left, right, and both:
https://github.com/sampsyo/wideq/blob/919139d9e81550009afe30dce2739b9fbf995799/wideq/ac.py#L113-L116

…which apparently is the right thing to do for some models but not yours.

I'm not sure exactly what to do here. I guess we need to query what type of model it is at setup and adjust the "on" state accordingly? That is, when we create the `ACDevice` object, it should check whether its `Operation` option supports `@AC_MAIN_OPERATION_ALL_ON_W` and, if so, switches the enum value for turning the device on from `RIGHT_ON` to `ALL_ON`.

Every time I interact with this LG API, we find another messy mistake…

**Comment by petarvucetin** [comment]
What a PITA.

I like your idea but I am afraid you would perpetually keep adding new mess :) This would require keeping a dictionary of models and their commands. These are pretty volatile going forward?

I would catch exception and try another command until it does not throw exception. This is "programming by exception" which is anti-pattern but their mess will propagate and can only be defended against if you introduce "anti-corruption" layer which I am not convinced is worth your time.

Another idea would be to specify in configuration (somehow) which one to use e.g. All, Left, Right? Let consumer pick the one that works.

**Comment by sampsyo** [comment]
Well, fortunately we wouldn't have to _literally_ hardcode a model->behavior mapping. We could just look in the model information dictionary to check which modes it supports. So at least it would generalize to similar models that have (or don't have) the left & right modes.

In the end, I think you end up sending the number (0 or 1). But we use the string key to avoid hard-coding the specific number for each option.

**Comment by ticky** [comment]
Thanks for merging my patch for this - I’d love to get a fresh release out with the fix included soon :smile:

**Comment by sampsyo** [comment]
Done! Just published v1.4.0.
